pkg/webhook/mutation.go

```diff
@@ -162,9 +170,56 @@ func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) adm
 }
 
 func (h *mutationHandler) mutateRequest(ctx context.Context, req admission.Request) (admission.Response, error) {
-	// TODO: place mutation logic here
-	mutatedObject := req.Object
-	resp := admission.PatchResponseFromRaw(req.Object.Raw, mutatedObject.Raw)
+
+	okResp := admission.Allowed("")
+	ns := &corev1.Namespace{}
+
+	// if the object being mutated is a namespace itself, we use it as namespace
+	if req.Kind.Kind == namespaceKind && req.Kind.Group == "" {
+		req.Namespace = ""
+		obj, _, err := deserializer.Decode(req.Object.Raw, nil, &corev1.Namespace{})
+		if err != nil {
+			return admission.Errored(int32(http.StatusInternalServerError), err), nil
+		}
+		ok := false
+		ns, ok = obj.(*corev1.Namespace)
+		if !ok {
+			return admission.Errored(int32(http.StatusInternalServerError), errors.New("failed to cast namespace object")), nil
+		}
+	} else {
+		if err := h.client.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				log.Info("Namespace not found", "name", req.AdmissionRequest.Namespace)
+				return okResp, nil
+			}
+			// bypass cached client and ask api-server directly
+			err = h.reader.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns)
+			if err != nil {
+				log.Info("Namespace not found", "name", req.AdmissionRequest.Namespace)
+				return okResp, nil
+			}
+		}
+	}
+	obj := unstructured.Unstructured{}
+	err := obj.UnmarshalJSON(req.Object.Raw)
+	if err != nil {
+		log.Error(err, "Failed to unmarshal", "object", string(req.Object.Raw))
+		return admission.Errored(int32(http.StatusInternalServerError), err), nil
+	}
+
+	err = h.mutationSystem.Mutate(&obj, ns)
+	if err != nil {
+		log.Error(err, "Failed to mutate object", "object", string(req.Object.Raw))
+		return admission.Errored(int32(http.StatusInternalServerError), err), nil
+	}
+
+	newJSON, err := obj.MarshalJSON()
+	if err != nil {
+		log.Error(err, "Failed to marshal mutated object", "object", obj)
+		return admission.Errored(int32(http.StatusInternalServerError), err), nil
+	}
+	log.Info("Mutated", "object", obj, "json", string(newJSON))
```

Comment:
Just noticed this, I think it is a debug leftover, will remove it.

Reply:
Good catch!